### ISSUE TITLE
Need Access to SpyStrategy plan

### ISSUE THREAD

**Issue by longmatthewh** [jasmine/jasmine#720]
I'm working on upgrading a jasmine helper called jasmine-stealth to support jasmine 2.x. One of the issues I'm having is accessing the SpyStrategy plan. Adding a function to expose plan would be helpful. Would you be interested in a pull request?

**Comment by jgrund** [comment]
One nice thing you could do in Jasmine 1.x was to reference nested spies without having to save a variable reference to them.

```
 var spy = jasmine.createSpy('foo').andReturn(jasmine.createSpy('bar');

spy.plan() // <-- bar spy
```

With 2.x the code becomes more verbose as you have to save off references to nested spies you'd like to inspect later.

**Comment by johnsonw** [comment]
When I discovered how to use .pan() I was able to go back into some of my tests and clean up a lot of unnecessary code. In particular, I was able to remove unnecessary variables that acted as spy permutations. As Joe mentioned above, using .plan() returns whatever you indicate your spy should return. Here's an example:

    var commandResult = {
        result: $q.when()
      };
    var command = jasmine.createSpy('command')
      .andReturn(commandResult);

    // command gets called in the code

    commandResult.result.then(function () {
        expect(...); // some expectation here
    });

Now imagine if you have to mock 10 commands, each with a different result. You need to create 10 command result variables. This can clutter things up quickly. Using .plan(), however, eliminates the need to do this:

    var command = jasmine.createSpy('command')
      .andReturn({
        result: $q.when()
      });

    // command gets called in the code

    command.plan().result.then(function () {
        expect(...); // some expectation here
    });

I feel like this simplifies the code and describes what you are doing, "I'm getting the result of the command spy" instead of going back and having to determine which result variable to use for each command.

**Comment by slackersoft** [comment]
I think I'm seeing two different issues here:

1. A desire to have the ability to provide a strategy to run based on the arguments that are passed, without having to `callFake` and write the `if` statement manually.
1. A desire to get the result of invoking a particular spy.

The first has been proposed before, and I think we'd be interested in seeing something like this get into jasmine-core, so I'd be happy to review a pull request for that.

The second seems like it might actually be solved with #474 which allows you to say: `command.calls.mostRecent().returnValue` to get the result.

**Comment by longmatthewh** [comment]
I ended up using sinon.js, but will work on a pull request in the near
future.

**Comment by GabiGrin** [comment]
Hey, I have the same request.
I'm trying to extend the spy's functionality so I can return a resolved/rejected promise with some value. For example:
`spyOn(obj, 'someMethod').andReturnResolvedPromise('response from server')`

With Jasmine 1 I just added to the Spy's prototype, but with 2+ the strategy is hidden.

Any news on allowing to extend spy's strategies?

**Comment by slackersoft** [comment]
We're looking at support for both custom strategy types for a suite as well as the ability to specify a strategy based on the arguments passed upon invocation. I'd be happy to review pull requests for either of these features if anyone has time before we have a chance to get to them. You can always see our current backlog in tracker here: https://www.pivotaltracker.com/n/projects/10606.

In the meantime, you should also be able to call the `exec` function on the `spyStrategy` to call whatever the current plan is.

**Comment by jgrund** [comment]
> The second seems like it might actually be solved with #474 which allows you to say: command.calls.mostRecent().returnValue to get the result.

That appears to be different. `returnValue` is the value returned by the spy. What I am looking for is access to the actual nested spy so I can assert calls on it directly.

**Comment by caseyhoward** [comment]
Since this doesn't seems to be going anywhere, what I'm wondering is if there a better way to stub a function with arguments besides using and.callFake and manually checking arguments. Also, making expectations after the fact is also out of the question for me since IMHO you should only be doing that for functions with side effects.

**Comment by slackersoft** [comment]
Jasmine 3.0 includes both the ability to add a custom spy strategy for your suite and the ability to specify a spy strategy specific to a list of arguments to the spy. I think these two features solve the underlying problems for this issue.

**Comment by slackersoft** [comment]
We haven't heard anything else on this issue and we believe the underlying use case should be solved in Jasmine 3.0, so I'm going to close this. Please open a new issue if there is a specific use case for spies that is still not solved in current releases of Jasmine.

Thanks for using Jasmine!
